provider/core/src/key.rs

```diff
@@ -463,213 +292,286 @@ impl DataKey {
         }
     }
 
-    /// Returns [`Ok`] if this data key matches the argument, or the appropriate error.
+    /// Gets the path as a static string slice.
+    #[inline]
+    pub const fn get(self) -> &'static str {
+        unsafe {
+            // Safe due to invariant that self.path is tagged correctly
+            core::str::from_utf8_unchecked(core::slice::from_raw_parts(
+                self.tagged.as_ptr().add(leading_tag!().len()),
+                self.tagged.len() - trailing_tag!().len() - leading_tag!().len(),
+            ))
+        }
+    }
+
+    /// Gets a platform-independent hash of a [`DataMarkerPath`].
+    ///
+    /// The hash is 4 bytes and allows for fast comparison.
     ///
-    /// Convenience method for data providers that support a single [`DataKey`].
+    /// # Example
+    ///
+    /// ```
+    /// use icu_provider::DataMarkerPath;
+    /// use icu_provider::DataMarkerPathHash;
+    ///
+    /// const PATH: DataMarkerPath = icu_provider::data_marker_path!("foo@1");
+    /// const PATH_HASH: DataMarkerPathHash = PATH.hashed();
+    ///
+    /// assert_eq!(PATH_HASH.to_bytes(), [0xe2, 0xb6, 0x17, 0x71]);
+    /// ```
+    #[inline]
+    pub const fn hashed(self) -> DataMarkerPathHash {
+        self.hash
+    }
+}
+
+impl Deref for DataMarkerPath {
+    type Target = str;
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.get()
+    }
+}
+
+/// Used for loading data from a dynamic ICU4X data provider.
+///
+/// A data marker is tightly coupled with the code that uses it to load data at runtime.
+/// Executables can be searched for `DataMarkerInfo` instances to produce optimized data files.
+/// Therefore, users should not generally create DataMarkerInfo instances; they should instead use
+/// the ones exported by a component.
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[allow(clippy::exhaustive_structs)] // TODO
+pub struct DataMarkerInfo {
```

Comment:
Issue: Should be non-exhaustive

(To keep it non-exhaustive, keep using the struct)